pkg/webui/components/qr/input/video/index.js

```diff
@@ -28,97 +28,165 @@ const m = defineMessages({
+    try {
+      if (!deviceId) {
         const userStream = await navigator.mediaDevices.getUserMedia({
-          video: videoMode ? { ...videoMode } : { facingMode: 'environment' },
+          video: { facingMode: 'environment' },
         })
+        // After requesting the stream, get the devices again as we now have permission to see all devices
+        getDevices()
+        // On initial request set the device id from the stream, this should rerender this component
+        setDeviceIdFromStream(userStream)
+      } else {
+        const userStream = await navigator.mediaDevices.getUserMedia({
+          video: { deviceId },
+        })
+        // Only set the stream if the device id is set
         setStream(userStream)
-      } catch (error) {
-        if (error instanceof DOMException && error.name === 'NotAllowedError') {
-          setCapture(false)
-          setError(true)
-        } else {
-          throw error
-        }
+      }
+    } catch (error) {
+      if (error instanceof DOMException && error.name === 'NotAllowedError') {
+        setCapture(false)
+        setError(true)
+      } else {
+        throw error
       }
     }
-  }, [devices, setCapture, setError, stream])
-
-  const switchStream = useCallback(async () => {
-    const ua = navigator.userAgent.toLowerCase()
-    if (ua.indexOf('safari') !== -1 && ua.indexOf('chrome') === -1) {
-      const userStream = await navigator.mediaDevices.getUserMedia({
-        video: {
-          facingMode: {
-            exact: videoMode.facingMode.exact === 'environment' ? 'user' : 'environment',
-          },
-        },
-      })
-      setStream(userStream)
-    } else if (videoMode.facingMode === 'environment') {
-      const userStream = await navigator.mediaDevices.getUserMedia({
-        video: { facingMode: 'user' },
-      })
-      setStream(userStream)
-    } else if (videoMode.facingMode === 'user') {
-      const userStream = await navigator.mediaDevices.getUserMedia({
-        video: { facingMode: 'environment' },
-      })
-      setStream(userStream)
-    } else if ('deviceId' in videoMode) {
-      let indexOfCurrentDevice = cameras.findIndex(camera => camera.deviceId === videoMode.deviceId)
-      // The first item will be taken from the beginning of the array after the last item.
-      const nextIndex = ++indexOfCurrentDevice % cameras.length
-      const device = cameras[nextIndex]
-      setVideoMode({ deviceId: device.deviceId })
-      const userStream = await navigator.mediaDevices.getUserMedia({
-        video: { deviceId: device.deviceId },
-      })
-      setStream(userStream)
-    }
-  }, [cameras, videoMode])
+  }, [deviceId, getDevices, setCapture, setDeviceIdFromStream, setError])
 
   useEffect(() => {
-    getDevices()
     getStream()
-    return () => {
-      if (stream) {
-        stream.getTracks().map(t => t.stop())
-      }
-    }
-  }, [devices, getDevices, getStream, stream])
+  }, [getStream])
```

Comment:
Just use `useEffect()` directly without wrapping around `useCallback()`.

Reply:
I don't think that will work because of the `async`

Reply:
```js
useEffect(() => {
  const asyncStuff = async () => {
    // Do async stuff
  }
  
  asyncStuff()
}, […])
```